File: 3743.maximize-cyclic-partition-score.java
#
# @lc app=leetcode id=3743 lang=java
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
class Solution {
    private static final int LOG = 11;

    @FunctionalInterface
    interface GetRangeFunc {
        long get(int l, int r);
    }

    private long getRange(int l, int r, long[][] stMax, long[][] stMin, int[] logg) {
        if (l >= r) return 0;
        int len = r - l;
        int lv = logg[len];
        int pw = 1 << lv;
        long mx = Math.max(stMax[lv][l], stMax[lv][r - pw]);
        long mn = Math.min(stMin[lv][l], stMin[lv][r - pw]);
        return mx - mn;
    }

    private void dcOpt(int L, int R, int optL, int optR, long[] prev, long[] curr, int s, GetRangeFunc gr) {
        if (L > R) return;
        int m = (L + R) / 2;
        long bestVal = Long.MIN_VALUE / 2;
        int bestK = -1;
        int startK = Math.max(optL, L - 1);
        int endK = Math.min(m - 1, optR);
        for (int kk = startK; kk <= endK; kk++) {
            long v = prev[kk] + gr.get(s + kk, s + m);
            if (v > bestVal) {
                bestVal = v;
                bestK = kk;
            }
        }
        curr[m] = bestVal;
        dcOpt(L, m - 1, optL, bestK, prev, curr, s, gr);
        dcOpt(m + 1, R, bestK, optR, prev, curr, s, gr);
    }

    public long maximumScore(int[] nums, int k) {
        int n = nums.length;
        int N = 2 * n;
        long[] A = new long[N];
        for (int i = 0; i < n; i++) {
            A[i] = nums[i];
            A[n + i] = nums[i];
        }
        long[][] stMax = new long[LOG][N];
        long[][] stMin = new long[LOG][N];
        for (int i = 0; i < N; i++) {
            stMax[0][i] = stMin[0][i] = A[i];
        }
        for (int lv = 1; lv < LOG; lv++) {
            for (int i = 0; i + (1 << lv) <= N; i++) {
                stMax[lv][i] = Math.max(stMax[lv - 1][i], stMax[lv - 1][i + (1 << (lv - 1))]);
                stMin[lv][i] = Math.min(stMin[lv - 1][i], stMin[lv - 1][i + (1 << (lv - 1))]);
            }
        }
        int[] logg = new int[N + 1];
        logg[1] = 0;
        for (int i = 2; i <= N; i++) {
            logg[i] = logg[i / 2] + 1;
        }
        long ans = 0;
        for (int s = 0; s < n; s++) {
            long[][] dp = new long[k + 1][n + 1];
            for (int len = 1; len <= n; len++) {
                dp[1][len] = getRange(s, s + len, stMax, stMin, logg);
            }
            ans = Math.max(ans, dp[1][n]);
            for (int p = 2; p <= k; p++) {
                GetRangeFunc gr = (ll, rr) -> getRange(ll, rr, stMax, stMin, logg);
                dcOpt(p, n, p - 1, n - 1, dp[p - 1], dp[p], s, gr);
                ans = Math.max(ans, dp[p][n]);
            }
        }
        return ans;
    }
}
# @lc code=end

File: 3743.maximize-cyclic-partition-score.go
#
# @lc app=leetcode id=3743 lang=golang
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
func maximumScore(nums []int, k int) int64 {
	n := len(nums)
	if n == 0 {
		return 0
	}
	const INF int64 = 1 << 60
	var globalMax, globalMin int64 = -INF, INF
	for _, v := range nums {
		iv := int64(v)
		if iv > globalMax {
			globalMax = iv
		}
		if iv < globalMin {
			globalMin = iv
		}
	}
	fullRange := globalMax - globalMin

	// prefix max/min [0..l-1]
	prefixMax := make([]int64, n+1)
	prefixMin := make([]int64, n+1)
	prefixMax[0] = -INF
	prefixMin[0] = INF
	cmx, cmn := int64(-INF), int64(INF)
	for l := 1; l <= n; l++ {
		cmx = max(cmx, int64(nums[l-1]))
		cmn = min(cmn, int64(nums[l-1]))
		prefixMax[l] = cmx
		prefixMin[l] = cmn
	}

	// suffix max/min last l elems nums[n-l : n]
	suffixMax := make([]int64, n+1)
	suffixMin := make([]int64, n+1)
	suffixMax[0] = -INF
	suffixMin[0] = INF
	cmx, cmn = int64(-INF), int64(INF)
	for l := 1; l <= n; l++ {
		idx := n - l
		cmx = max(cmx, int64(nums[idx]))
		cmn = min(cmn, int64(nums[idx]))
		suffixMax[l] = cmx
		suffixMin[l] = cmn
	}

	// range_max[start][leng] = max(nums[start:start+leng])
	rangeMaxs := make([][]int64, n)
	rangeMins := make([][]int64, n)
	for start := 0; start < n; start++ {
		maxLen := n - start
		rangeMaxs[start] = make([]int64, maxLen+1)
		rangeMins[start] = make([]int64, maxLen+1)
		cmx = int64(-INF)
		cmn = int64(INF)
		for leng := 1; leng <= maxLen; leng++ {
			cmx = max(cmx, int64(nums[start+leng-1]))
			cmn = min(cmn, int64(nums[start+leng-1]))
			rangeMaxs[start][leng] = cmx
			rangeMins[start][leng] = cmn
		}
	}

	dp := make([][]int64, k+1)
	for j := 0; j <= k; j++ {
		dp[j] = make([]int64, n)
		for ii := 0; ii < n; ii++ {
			dp[j][ii] = -INF
		}
	}
	for i := 0; i < n; i++ {
		dp[1][i] = fullRange
	}

	for j := 2; j <= k; j++ {
		for i := 0; i < n; i++ {
			dp[j][i] = dp[j-1][i]
			for lenf := 1; lenf < n; lenf++ {
				nexti := (i + lenf) % n
				var c int64
				if i+lenf <= n {
					c = rangeMaxs[i][lenf] - rangeMins[i][lenf]
				} else {
					suflen := n - i
					prelen := lenf - suflen
					pmx := prefixMax[prelen]
					pmn := prefixMin[prelen]
					smx := suffixMax[suflen]
					smn := suffixMin[suflen]
					c = max(pmx, smx) - min(pmn, smn)
				}
				if dp[j-1][nexti] != -INF {
					dp[j][i] = max(dp[j][i], c+dp[j-1][nexti])
				}
			}
		}
	}

	ans := int64(0)
	for j := 1; j <= k; j++ {
		for i := 0; i < n; i++ {
			ans = max(ans, dp[j][i])
		}
	}
	return ans
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}
# @lc code=end